{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f60386b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pyDEseq2 requires:\n",
    "# - numpy==1.23.0\n",
    "# - anndata==0.8.0\n",
    "# - pandas==1.4.3\n",
    "# - scikit-learn==1.1.1\n",
    "# - scipy==1.8.1\n",
    "# - statsmodels==0.13.2\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import matplotlib.pyplot as plt\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "from subprocess import Popen, PIPE\n",
    "from math import log\n",
    "import pyensembl as ensembl\n",
    "import io\n",
    "ensembldb = ensembl.EnsemblRelease(release = 86)\n",
    "\n",
    "genes = ensembldb.genes()\n",
    "coding_genes = [gene for gene in genes if gene.biotype == 'protein_coding']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "84db9af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup Metadata\n",
    "samples = {'P1_T':'../data/SN_MO/P1_T_outs/P1_T_outs_atac_fragments.tsv.gz',\n",
    "           'P1_N':'../data/SN_MO/P1_N_outs/P1_N_outs_atac_fragments.tsv.gz',\n",
    "           'P2_T':'../data/SN_MO/P2_T_outs/P2_T_outs_atac_fragments.tsv.gz',\n",
    "           'P2_N':'../data/SN_MO/P2_N_outs/P2_N_outs_atac_fragments.tsv.gz',\n",
    "           'P3_T':'../data/SN_MO/P3_T_outs/P3_T_outs_atac_fragments.tsv.gz',\n",
    "           'P4_T':'../data/SN_MO/P4_T_outs/P4_T_outs_atac_fragments.tsv.gz'}\n",
    "\n",
    "barcodes = {}\n",
    "\n",
    "for sample in samples.keys():\n",
    "    with open('../results/%s/%s_cort_barcodes_formatted.csv' % (sample, sample), 'r') as fin:\n",
    "        barcodes[sample] = [line.strip() for line in fin]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "40438e4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PeakID</th>\n",
       "      <th>Query</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_1</th>\n",
       "      <td>cort_peak_1</td>\n",
       "      <td>GL000008.2:2217-2483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_2</th>\n",
       "      <td>cort_peak_2</td>\n",
       "      <td>GL000008.2:2531-3604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_3</th>\n",
       "      <td>cort_peak_3</td>\n",
       "      <td>GL000008.2:3731-3881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_4</th>\n",
       "      <td>cort_peak_4</td>\n",
       "      <td>GL000008.2:83282-83452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_5</th>\n",
       "      <td>cort_peak_5</td>\n",
       "      <td>GL000008.2:83571-83972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181882</th>\n",
       "      <td>cort_peak_181882</td>\n",
       "      <td>chrY:56843616-56843854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181883</th>\n",
       "      <td>cort_peak_181883</td>\n",
       "      <td>chrY:56849846-56849939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181884</th>\n",
       "      <td>cort_peak_181884</td>\n",
       "      <td>chrY:56850358-56850506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181885</th>\n",
       "      <td>cort_peak_181885</td>\n",
       "      <td>chrY:56850763-56850990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181886</th>\n",
       "      <td>cort_peak_181886</td>\n",
       "      <td>chrY:56851298-56851687</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181886 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            PeakID                   Query\n",
       "PeakID                                                    \n",
       "cort_peak_1            cort_peak_1    GL000008.2:2217-2483\n",
       "cort_peak_2            cort_peak_2    GL000008.2:2531-3604\n",
       "cort_peak_3            cort_peak_3    GL000008.2:3731-3881\n",
       "cort_peak_4            cort_peak_4  GL000008.2:83282-83452\n",
       "cort_peak_5            cort_peak_5  GL000008.2:83571-83972\n",
       "...                            ...                     ...\n",
       "cort_peak_181882  cort_peak_181882  chrY:56843616-56843854\n",
       "cort_peak_181883  cort_peak_181883  chrY:56849846-56849939\n",
       "cort_peak_181884  cort_peak_181884  chrY:56850358-56850506\n",
       "cort_peak_181885  cort_peak_181885  chrY:56850763-56850990\n",
       "cort_peak_181886  cort_peak_181886  chrY:56851298-56851687\n",
       "\n",
       "[181886 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = pd.read_csv('../results/ATAC/macs/cort_peaks.narrowPeak', sep='\\t', header=None)\n",
    "peaks = peaks[[0,1,2,3]]\n",
    "peaks.columns = ['Chr', 'Start', 'End', 'PeakID']\n",
    "peaks.index = peaks['PeakID']\n",
    "peak_ids = peaks['PeakID']\n",
    "queries = peak_ids.map(lambda x: '{}:{}-{}'.format(peaks.loc[x]['Chr'], peaks.loc[x]['Start'], peaks.loc[x]['End']))\n",
    "queries = queries.rename('Query')\n",
    "peak_key = pd.concat([peak_ids,queries], axis=1)\n",
    "peak_key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7e300e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Assemble Counts DF using row-wise apply function, adding columns. \n",
    "# Function goes row by row and uses the query built before and the tabix command to query the fragments file.\n",
    "# Fragments are then filtered based on the corticotroph barcodes\n",
    "\n",
    "\n",
    "def f(x):\n",
    "    result = []\n",
    "    for sample in samples.keys():\n",
    "        process = Popen(['tabix', samples[sample], x['Query']], stdout=PIPE, universal_newlines=True)\n",
    "        process = process.communicate()\n",
    "        if process[0] == '':\n",
    "            result.append(0)\n",
    "        elif process[0] != '':\n",
    "            frags_df = pd.read_csv(io.StringIO(process[0]), sep = '\\t',header=None).drop([4], axis=1)\n",
    "            frags_df = frags_df.loc[(frags_df[3].isin(barcodes[sample]))]\n",
    "            result.append(len(frags_df))\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "934cc3bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>P1_T</th>\n",
       "      <th>P1_N</th>\n",
       "      <th>P2_T</th>\n",
       "      <th>P2_N</th>\n",
       "      <th>P3_T</th>\n",
       "      <th>P4_T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_1</th>\n",
       "      <td>GL000008.2:2217-2483</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_2</th>\n",
       "      <td>GL000008.2:2531-3604</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_3</th>\n",
       "      <td>GL000008.2:3731-3881</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_4</th>\n",
       "      <td>GL000008.2:83282-83452</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_5</th>\n",
       "      <td>GL000008.2:83571-83972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181882</th>\n",
       "      <td>chrY:56843616-56843854</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181883</th>\n",
       "      <td>chrY:56849846-56849939</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181884</th>\n",
       "      <td>chrY:56850358-56850506</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181885</th>\n",
       "      <td>chrY:56850763-56850990</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181886</th>\n",
       "      <td>chrY:56851298-56851687</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>700</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181886 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Query  P1_T  P1_N  P2_T  P2_N  P3_T  P4_T\n",
       "PeakID                                                                      \n",
       "cort_peak_1         GL000008.2:2217-2483     0     0     0     0     0     0\n",
       "cort_peak_2         GL000008.2:2531-3604     0     0     0     0     0     0\n",
       "cort_peak_3         GL000008.2:3731-3881     0     0     0     0     0     0\n",
       "cort_peak_4       GL000008.2:83282-83452     0     0     0     0     0     0\n",
       "cort_peak_5       GL000008.2:83571-83972     0     0     0     0     0     0\n",
       "...                                  ...   ...   ...   ...   ...   ...   ...\n",
       "cort_peak_181882  chrY:56843616-56843854     1     5     2     0    75     6\n",
       "cort_peak_181883  chrY:56849846-56849939     2     0     1     0     9     3\n",
       "cort_peak_181884  chrY:56850358-56850506     1     3     0     0    22    11\n",
       "cort_peak_181885  chrY:56850763-56850990     1     3     1     2    38    17\n",
       "cort_peak_181886  chrY:56851298-56851687     1     5     4     3   700    11\n",
       "\n",
       "[181886 rows x 7 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = pd.DataFrame(peak_key['Query'])\n",
    "          \n",
    "counts[list(samples.keys())] = counts.apply(f, axis=1, result_type='expand')\n",
    "\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9fd814f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.to_csv('../results/ATAC/differential/peak_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a31a73e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1_T</th>\n",
       "      <th>P1_N</th>\n",
       "      <th>P2_T</th>\n",
       "      <th>P2_N</th>\n",
       "      <th>P3_T</th>\n",
       "      <th>P4_T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181882</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181883</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181884</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181885</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181886</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>700</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181886 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  P1_T  P1_N  P2_T  P2_N  P3_T  P4_T\n",
       "PeakID                                              \n",
       "cort_peak_1          0     0     0     0     0     0\n",
       "cort_peak_2          0     0     0     0     0     0\n",
       "cort_peak_3          0     0     0     0     0     0\n",
       "cort_peak_4          0     0     0     0     0     0\n",
       "cort_peak_5          0     0     0     0     0     0\n",
       "...                ...   ...   ...   ...   ...   ...\n",
       "cort_peak_181882     1     5     2     0    75     6\n",
       "cort_peak_181883     2     0     1     0     9     3\n",
       "cort_peak_181884     1     3     0     0    22    11\n",
       "cort_peak_181885     1     3     1     2    38    17\n",
       "cort_peak_181886     1     5     4     3   700    11\n",
       "\n",
       "[181886 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = pd.read_csv('../results/ATAC/differential/peak_counts.csv', index_col=0)\n",
    "counts = counts.drop('Query', axis=1)\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e92c39ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>cort_peak_11</th>\n",
       "      <th>cort_peak_14</th>\n",
       "      <th>cort_peak_15</th>\n",
       "      <th>cort_peak_16</th>\n",
       "      <th>cort_peak_17</th>\n",
       "      <th>cort_peak_19</th>\n",
       "      <th>cort_peak_21</th>\n",
       "      <th>cort_peak_22</th>\n",
       "      <th>cort_peak_23</th>\n",
       "      <th>cort_peak_25</th>\n",
       "      <th>...</th>\n",
       "      <th>cort_peak_181877</th>\n",
       "      <th>cort_peak_181878</th>\n",
       "      <th>cort_peak_181879</th>\n",
       "      <th>cort_peak_181880</th>\n",
       "      <th>cort_peak_181881</th>\n",
       "      <th>cort_peak_181882</th>\n",
       "      <th>cort_peak_181883</th>\n",
       "      <th>cort_peak_181884</th>\n",
       "      <th>cort_peak_181885</th>\n",
       "      <th>cort_peak_181886</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_T</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_N</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_T</th>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_N</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_T</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>26</td>\n",
       "      <td>43</td>\n",
       "      <td>32</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>38</td>\n",
       "      <td>700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_T</th>\n",
       "      <td>100</td>\n",
       "      <td>24</td>\n",
       "      <td>49</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>47</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>62</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 175624 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID  cort_peak_11  cort_peak_14  cort_peak_15  cort_peak_16  cort_peak_17  \\\n",
       "P1_T               0             0             0             0             0   \n",
       "P1_N               2             2             6             5             0   \n",
       "P2_T              21             1             2             2             3   \n",
       "P2_N               0             0             0             0             0   \n",
       "P3_T               4             0             8             4             8   \n",
       "P4_T             100            24            49            15            15   \n",
       "\n",
       "PeakID  cort_peak_19  cort_peak_21  cort_peak_22  cort_peak_23  cort_peak_25  \\\n",
       "P1_T               2             0             0             1             0   \n",
       "P1_N              19             1             1             8             1   \n",
       "P2_T              29             1             1            11             0   \n",
       "P2_N               1             1             0             0             1   \n",
       "P3_T               7             2             0             3            26   \n",
       "P4_T              47            16            16            62            17   \n",
       "\n",
       "PeakID  ...  cort_peak_181877  cort_peak_181878  cort_peak_181879  \\\n",
       "P1_T    ...                 1                 2                 4   \n",
       "P1_N    ...                 2                 2                 4   \n",
       "P2_T    ...                 2                 3                 3   \n",
       "P2_N    ...                 1                 1                 0   \n",
       "P3_T    ...                16                21                26   \n",
       "P4_T    ...                 7                 8                10   \n",
       "\n",
       "PeakID  cort_peak_181880  cort_peak_181881  cort_peak_181882  \\\n",
       "P1_T                   0                 1                 1   \n",
       "P1_N                   3                 0                 5   \n",
       "P2_T                   5                 3                 2   \n",
       "P2_N                   0                 1                 0   \n",
       "P3_T                  43                32                75   \n",
       "P4_T                   2                 9                 6   \n",
       "\n",
       "PeakID  cort_peak_181883  cort_peak_181884  cort_peak_181885  cort_peak_181886  \n",
       "P1_T                   2                 1                 1                 1  \n",
       "P1_N                   0                 3                 3                 5  \n",
       "P2_T                   1                 0                 1                 4  \n",
       "P2_N                   0                 0                 2                 3  \n",
       "P3_T                   9                22                38               700  \n",
       "P4_T                   3                11                17                11  \n",
       "\n",
       "[6 rows x 175624 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = counts.transpose()\n",
    "# Filter out features with low fragment counts\n",
    "genes_to_keep = counts.columns[counts.sum(axis=0) >= 10]\n",
    "counts = counts[genes_to_keep]\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "43b19362",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>P1_T</th>\n",
       "      <th>P1_N</th>\n",
       "      <th>P2_T</th>\n",
       "      <th>P2_N</th>\n",
       "      <th>P3_T</th>\n",
       "      <th>P4_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>chrX:100622109-100644991</td>\n",
       "      <td>21</td>\n",
       "      <td>43</td>\n",
       "      <td>97</td>\n",
       "      <td>16</td>\n",
       "      <td>951</td>\n",
       "      <td>573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>chrX:100579802-100604885</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>231</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>chr20:50929867-50963555</td>\n",
       "      <td>51</td>\n",
       "      <td>153</td>\n",
       "      <td>334</td>\n",
       "      <td>37</td>\n",
       "      <td>4398</td>\n",
       "      <td>1073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>chr1:169844631-169899267</td>\n",
       "      <td>60</td>\n",
       "      <td>131</td>\n",
       "      <td>414</td>\n",
       "      <td>35</td>\n",
       "      <td>4199</td>\n",
       "      <td>1053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>chr1:169657007-169859080</td>\n",
       "      <td>105</td>\n",
       "      <td>152</td>\n",
       "      <td>298</td>\n",
       "      <td>92</td>\n",
       "      <td>2929</td>\n",
       "      <td>756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-210D15.9</th>\n",
       "      <td>chrX:135304480-135314659</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>117</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-244E17.1</th>\n",
       "      <td>chr14:23094065-23109992</td>\n",
       "      <td>71</td>\n",
       "      <td>231</td>\n",
       "      <td>425</td>\n",
       "      <td>41</td>\n",
       "      <td>5005</td>\n",
       "      <td>1447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTB-60B18.23</th>\n",
       "      <td>chr19:49050793-49070076</td>\n",
       "      <td>29</td>\n",
       "      <td>60</td>\n",
       "      <td>114</td>\n",
       "      <td>16</td>\n",
       "      <td>897</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP3-454G6.2</th>\n",
       "      <td>chr1:171595621-171643799</td>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>116</td>\n",
       "      <td>21</td>\n",
       "      <td>644</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP5-937E21.8</th>\n",
       "      <td>chrX:149543210-149554852</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19933 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Query  P1_T  P1_N  P2_T  P2_N  P3_T  P4_T\n",
       "TSPAN6         chrX:100622109-100644991    21    43    97    16   951   573\n",
       "TNMD           chrX:100579802-100604885    12    18    30     8   231    99\n",
       "DPM1            chr20:50929867-50963555    51   153   334    37  4398  1073\n",
       "SCYL3          chr1:169844631-169899267    60   131   414    35  4199  1053\n",
       "C1orf112       chr1:169657007-169859080   105   152   298    92  2929   756\n",
       "...                                 ...   ...   ...   ...   ...   ...   ...\n",
       "RP13-210D15.9  chrX:135304480-135314659     2     6    11     3   117    30\n",
       "RP11-244E17.1   chr14:23094065-23109992    71   231   425    41  5005  1447\n",
       "CTB-60B18.23    chr19:49050793-49070076    29    60   114    16   897   236\n",
       "RP3-454G6.2    chr1:171595621-171643799    31    43   116    21   644   341\n",
       "RP5-937E21.8   chrX:149543210-149554852     4     5     4     0    42    25\n",
       "\n",
       "[19933 rows x 7 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_queries = {}\n",
    "for gene in coding_genes:\n",
    "    if gene.start < gene.end:\n",
    "        gene_queries[gene.gene_name] = 'chr%s:%d-%d' % (gene.contig, gene.start-5000, gene.end+5000)\n",
    "    elif gene.start > gene.end:\n",
    "        gene_queries[gene.gene_name] = 'chr%s:%d-%d' % (gene.contig, gene.start+5000, gene.end-5000)    \n",
    "gene_counts = pd.DataFrame(list(gene_queries.values()), index=gene_queries.keys(), columns=['Query'])\n",
    "gene_counts[list(samples.keys())] = gene_counts.apply(f, axis=1, result_type='expand')\n",
    "gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d6648669",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_counts.to_csv('../results/ATAC/differential/gene_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "554b1309",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1_T</th>\n",
       "      <th>P1_N</th>\n",
       "      <th>P2_T</th>\n",
       "      <th>P2_N</th>\n",
       "      <th>P3_T</th>\n",
       "      <th>P4_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>21</td>\n",
       "      <td>43</td>\n",
       "      <td>97</td>\n",
       "      <td>16</td>\n",
       "      <td>951</td>\n",
       "      <td>573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>231</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>51</td>\n",
       "      <td>153</td>\n",
       "      <td>334</td>\n",
       "      <td>37</td>\n",
       "      <td>4398</td>\n",
       "      <td>1073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>60</td>\n",
       "      <td>131</td>\n",
       "      <td>414</td>\n",
       "      <td>35</td>\n",
       "      <td>4199</td>\n",
       "      <td>1053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>105</td>\n",
       "      <td>152</td>\n",
       "      <td>298</td>\n",
       "      <td>92</td>\n",
       "      <td>2929</td>\n",
       "      <td>756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-210D15.9</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>117</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-244E17.1</th>\n",
       "      <td>71</td>\n",
       "      <td>231</td>\n",
       "      <td>425</td>\n",
       "      <td>41</td>\n",
       "      <td>5005</td>\n",
       "      <td>1447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTB-60B18.23</th>\n",
       "      <td>29</td>\n",
       "      <td>60</td>\n",
       "      <td>114</td>\n",
       "      <td>16</td>\n",
       "      <td>897</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP3-454G6.2</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>116</td>\n",
       "      <td>21</td>\n",
       "      <td>644</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP5-937E21.8</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19933 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               P1_T  P1_N  P2_T  P2_N  P3_T  P4_T\n",
       "TSPAN6           21    43    97    16   951   573\n",
       "TNMD             12    18    30     8   231    99\n",
       "DPM1             51   153   334    37  4398  1073\n",
       "SCYL3            60   131   414    35  4199  1053\n",
       "C1orf112        105   152   298    92  2929   756\n",
       "...             ...   ...   ...   ...   ...   ...\n",
       "RP13-210D15.9     2     6    11     3   117    30\n",
       "RP11-244E17.1    71   231   425    41  5005  1447\n",
       "CTB-60B18.23     29    60   114    16   897   236\n",
       "RP3-454G6.2      31    43   116    21   644   341\n",
       "RP5-937E21.8      4     5     4     0    42    25\n",
       "\n",
       "[19933 rows x 6 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_counts = gene_counts.drop('Query', axis=1)\n",
    "gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "58864954",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>RP5-994D16.12</th>\n",
       "      <th>RP11-402P6.15</th>\n",
       "      <th>AC074141.1</th>\n",
       "      <th>AL022397.1</th>\n",
       "      <th>AL022067.1</th>\n",
       "      <th>RP13-210D15.9</th>\n",
       "      <th>RP11-244E17.1</th>\n",
       "      <th>CTB-60B18.23</th>\n",
       "      <th>RP3-454G6.2</th>\n",
       "      <th>RP5-937E21.8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_T</th>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>60</td>\n",
       "      <td>105</td>\n",
       "      <td>45</td>\n",
       "      <td>39</td>\n",
       "      <td>22</td>\n",
       "      <td>129</td>\n",
       "      <td>78</td>\n",
       "      <td>...</td>\n",
       "      <td>62</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>122</td>\n",
       "      <td>60</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>29</td>\n",
       "      <td>31</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_N</th>\n",
       "      <td>43</td>\n",
       "      <td>18</td>\n",
       "      <td>153</td>\n",
       "      <td>131</td>\n",
       "      <td>152</td>\n",
       "      <td>69</td>\n",
       "      <td>66</td>\n",
       "      <td>54</td>\n",
       "      <td>281</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>196</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>298</td>\n",
       "      <td>139</td>\n",
       "      <td>6</td>\n",
       "      <td>231</td>\n",
       "      <td>60</td>\n",
       "      <td>43</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_T</th>\n",
       "      <td>97</td>\n",
       "      <td>30</td>\n",
       "      <td>334</td>\n",
       "      <td>414</td>\n",
       "      <td>298</td>\n",
       "      <td>88</td>\n",
       "      <td>86</td>\n",
       "      <td>120</td>\n",
       "      <td>544</td>\n",
       "      <td>530</td>\n",
       "      <td>...</td>\n",
       "      <td>329</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>890</td>\n",
       "      <td>269</td>\n",
       "      <td>11</td>\n",
       "      <td>425</td>\n",
       "      <td>114</td>\n",
       "      <td>116</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_N</th>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>37</td>\n",
       "      <td>35</td>\n",
       "      <td>92</td>\n",
       "      <td>17</td>\n",
       "      <td>31</td>\n",
       "      <td>25</td>\n",
       "      <td>96</td>\n",
       "      <td>62</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>109</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>41</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_T</th>\n",
       "      <td>951</td>\n",
       "      <td>231</td>\n",
       "      <td>4398</td>\n",
       "      <td>4199</td>\n",
       "      <td>2929</td>\n",
       "      <td>880</td>\n",
       "      <td>712</td>\n",
       "      <td>1291</td>\n",
       "      <td>7192</td>\n",
       "      <td>8249</td>\n",
       "      <td>...</td>\n",
       "      <td>4791</td>\n",
       "      <td>352</td>\n",
       "      <td>110</td>\n",
       "      <td>7712</td>\n",
       "      <td>1814</td>\n",
       "      <td>117</td>\n",
       "      <td>5005</td>\n",
       "      <td>897</td>\n",
       "      <td>644</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_T</th>\n",
       "      <td>573</td>\n",
       "      <td>99</td>\n",
       "      <td>1073</td>\n",
       "      <td>1053</td>\n",
       "      <td>756</td>\n",
       "      <td>208</td>\n",
       "      <td>222</td>\n",
       "      <td>400</td>\n",
       "      <td>1524</td>\n",
       "      <td>2387</td>\n",
       "      <td>...</td>\n",
       "      <td>795</td>\n",
       "      <td>117</td>\n",
       "      <td>74</td>\n",
       "      <td>2127</td>\n",
       "      <td>751</td>\n",
       "      <td>30</td>\n",
       "      <td>1447</td>\n",
       "      <td>236</td>\n",
       "      <td>341</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 19683 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      TSPAN6  TNMD  DPM1  SCYL3  C1orf112  FGR  CFH  FUCA2  GCLC  NFYA  ...  \\\n",
       "P1_T      21    12    51     60       105   45   39     22   129    78  ...   \n",
       "P1_N      43    18   153    131       152   69   66     54   281   290  ...   \n",
       "P2_T      97    30   334    414       298   88   86    120   544   530  ...   \n",
       "P2_N      16     8    37     35        92   17   31     25    96    62  ...   \n",
       "P3_T     951   231  4398   4199      2929  880  712   1291  7192  8249  ...   \n",
       "P4_T     573    99  1073   1053       756  208  222    400  1524  2387  ...   \n",
       "\n",
       "      RP5-994D16.12  RP11-402P6.15  AC074141.1  AL022397.1  AL022067.1  \\\n",
       "P1_T             62             15           6         122          60   \n",
       "P1_N            196             20           7         298         139   \n",
       "P2_T            329             22          22         890         269   \n",
       "P2_N             40              7           6         109          47   \n",
       "P3_T           4791            352         110        7712        1814   \n",
       "P4_T            795            117          74        2127         751   \n",
       "\n",
       "      RP13-210D15.9  RP11-244E17.1  CTB-60B18.23  RP3-454G6.2  RP5-937E21.8  \n",
       "P1_T              2             71            29           31             4  \n",
       "P1_N              6            231            60           43             5  \n",
       "P2_T             11            425           114          116             4  \n",
       "P2_N              3             41            16           21             0  \n",
       "P3_T            117           5005           897          644            42  \n",
       "P4_T             30           1447           236          341            25  \n",
       "\n",
       "[6 rows x 19683 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_counts = gene_counts.transpose()\n",
    "# Filter out features with low fragment counts\n",
    "genes_to_keep = gene_counts.columns[gene_counts.sum(axis=0) >= 10]\n",
    "gene_counts = gene_counts[genes_to_keep]\n",
    "gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7480ab0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_T</th>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_N</th>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_T</th>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_N</th>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_T</th>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_T</th>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Annotation\n",
       "P1_T      Tumor\n",
       "P1_N     Normal\n",
       "P2_T      Tumor\n",
       "P2_N     Normal\n",
       "P3_T      Tumor\n",
       "P4_T      Tumor"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df = pd.DataFrame(['Tumor', 'Normal', 'Tumor', 'Normal', 'Tumor', 'Tumor'], ['P1_T', 'P1_N', 'P2_T', 'P2_N', 'P3_T', 'P4_T'], [\"Annotation\"])\n",
    "clinical_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f0cc03be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6 × 175624\n",
       "    obs: 'Annotation'\n",
       "    obsm: 'design_matrix'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Peaks\n",
    "dds = DeseqDataSet(\n",
    "    counts=counts,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=\"Annotation\",\n",
    "    refit_cooks=True,\n",
    "    n_cpus=1,\n",
    ")\n",
    "dds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "920017b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[51], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m dds\u001b[38;5;241m.\u001b[39mdeseq2()\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/pydeseq2/dds.py:307\u001b[0m, in \u001b[0;36mDeseqDataSet.deseq2\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfit_size_factors()\n\u001b[1;32m    306\u001b[0m \u001b[38;5;66;03m# Fit an independent negative binomial model per gene\u001b[39;00m\n\u001b[0;32m--> 307\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfit_genewise_dispersions()\n\u001b[1;32m    308\u001b[0m \u001b[38;5;66;03m# Fit a parameterized trend curve for dispersions, of the form\u001b[39;00m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;66;03m# f(\\mu) = \\alpha_1/\\mu + a_0\u001b[39;00m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfit_dispersion_trend()\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/pydeseq2/dds.py:432\u001b[0m, in \u001b[0;36mDeseqDataSet.fit_genewise_dispersions\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    430\u001b[0m start \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m    431\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parallel_backend(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mloky\u001b[39m\u001b[38;5;124m\"\u001b[39m, inner_max_num_threads\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m):\n\u001b[0;32m--> 432\u001b[0m     res \u001b[38;5;241m=\u001b[39m Parallel(\n\u001b[1;32m    433\u001b[0m         n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_processes,\n\u001b[1;32m    434\u001b[0m         verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mjoblib_verbosity,\n\u001b[1;32m    435\u001b[0m         batch_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbatch_size,\n\u001b[1;32m    436\u001b[0m     )(\n\u001b[1;32m    437\u001b[0m         delayed(fit_alpha_mle)(\n\u001b[1;32m    438\u001b[0m             counts\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mX[:, i],\n\u001b[1;32m    439\u001b[0m             design_matrix\u001b[38;5;241m=\u001b[39mdesign_matrix,\n\u001b[1;32m    440\u001b[0m             mu\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlayers[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_mu_hat\u001b[39m\u001b[38;5;124m\"\u001b[39m][:, i],\n\u001b[1;32m    441\u001b[0m             alpha_hat\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvarm[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_rough_dispersions\u001b[39m\u001b[38;5;124m\"\u001b[39m][i],\n\u001b[1;32m    442\u001b[0m             min_disp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmin_disp,\n\u001b[1;32m    443\u001b[0m             max_disp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmax_disp,\n\u001b[1;32m    444\u001b[0m         )\n\u001b[1;32m    445\u001b[0m         \u001b[38;5;66;03m# for i in range(num_genes)\u001b[39;00m\n\u001b[1;32m    446\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnon_zero_idx\n\u001b[1;32m    447\u001b[0m     )\n\u001b[1;32m    448\u001b[0m end \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m    449\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m... done in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mend\u001b[38;5;250m \u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;250m \u001b[39mstart\u001b[38;5;132;01m:\u001b[39;00m\u001b[38;5;124m.2f\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m seconds.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/parallel.py:1088\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1085\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdispatch_one_batch(iterator):\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_iterating \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_original_iterator \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1088\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdispatch_one_batch(iterator):\n\u001b[1;32m   1089\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m   1091\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pre_dispatch \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m n_jobs \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   1092\u001b[0m     \u001b[38;5;66;03m# The iterable was consumed all at once by the above for loop.\u001b[39;00m\n\u001b[1;32m   1093\u001b[0m     \u001b[38;5;66;03m# No need to wait for async callbacks to trigger to\u001b[39;00m\n\u001b[1;32m   1094\u001b[0m     \u001b[38;5;66;03m# consumption.\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/parallel.py:901\u001b[0m, in \u001b[0;36mParallel.dispatch_one_batch\u001b[0;34m(self, iterator)\u001b[0m\n\u001b[1;32m    899\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    900\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 901\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_dispatch(tasks)\n\u001b[1;32m    902\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/parallel.py:819\u001b[0m, in \u001b[0;36mParallel._dispatch\u001b[0;34m(self, batch)\u001b[0m\n\u001b[1;32m    817\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lock:\n\u001b[1;32m    818\u001b[0m     job_idx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs)\n\u001b[0;32m--> 819\u001b[0m     job \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend\u001b[38;5;241m.\u001b[39mapply_async(batch, callback\u001b[38;5;241m=\u001b[39mcb)\n\u001b[1;32m    820\u001b[0m     \u001b[38;5;66;03m# A job can complete so quickly than its callback is\u001b[39;00m\n\u001b[1;32m    821\u001b[0m     \u001b[38;5;66;03m# called before we get here, causing self._jobs to\u001b[39;00m\n\u001b[1;32m    822\u001b[0m     \u001b[38;5;66;03m# grow. To ensure correct results ordering, .insert is\u001b[39;00m\n\u001b[1;32m    823\u001b[0m     \u001b[38;5;66;03m# used (rather than .append) in the following line\u001b[39;00m\n\u001b[1;32m    824\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs\u001b[38;5;241m.\u001b[39minsert(job_idx, job)\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/_parallel_backends.py:208\u001b[0m, in \u001b[0;36mSequentialBackend.apply_async\u001b[0;34m(self, func, callback)\u001b[0m\n\u001b[1;32m    206\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_async\u001b[39m(\u001b[38;5;28mself\u001b[39m, func, callback\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    207\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Schedule a func to be run\"\"\"\u001b[39;00m\n\u001b[0;32m--> 208\u001b[0m     result \u001b[38;5;241m=\u001b[39m ImmediateResult(func)\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m callback:\n\u001b[1;32m    210\u001b[0m         callback(result)\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/_parallel_backends.py:597\u001b[0m, in \u001b[0;36mImmediateResult.__init__\u001b[0;34m(self, batch)\u001b[0m\n\u001b[1;32m    594\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, batch):\n\u001b[1;32m    595\u001b[0m     \u001b[38;5;66;03m# Don't delay the application, to avoid keeping the input\u001b[39;00m\n\u001b[1;32m    596\u001b[0m     \u001b[38;5;66;03m# arguments in memory\u001b[39;00m\n\u001b[0;32m--> 597\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mresults \u001b[38;5;241m=\u001b[39m batch()\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/parallel.py:288\u001b[0m, in \u001b[0;36mBatchedCalls.__call__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    284\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    285\u001b[0m     \u001b[38;5;66;03m# Set the default nested backend to self._backend but do not set the\u001b[39;00m\n\u001b[1;32m    286\u001b[0m     \u001b[38;5;66;03m# change the default number of processes to -1\u001b[39;00m\n\u001b[1;32m    287\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m parallel_backend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend, n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_n_jobs):\n\u001b[0;32m--> 288\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m [func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    289\u001b[0m                 \u001b[38;5;28;01mfor\u001b[39;00m func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mitems]\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/joblib/parallel.py:288\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    284\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    285\u001b[0m     \u001b[38;5;66;03m# Set the default nested backend to self._backend but do not set the\u001b[39;00m\n\u001b[1;32m    286\u001b[0m     \u001b[38;5;66;03m# change the default number of processes to -1\u001b[39;00m\n\u001b[1;32m    287\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m parallel_backend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend, n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_n_jobs):\n\u001b[0;32m--> 288\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m [func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    289\u001b[0m                 \u001b[38;5;28;01mfor\u001b[39;00m func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mitems]\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/pydeseq2/utils.py:646\u001b[0m, in \u001b[0;36mfit_alpha_mle\u001b[0;34m(counts, design_matrix, mu, alpha_hat, min_disp, max_disp, prior_disp_var, cr_reg, prior_reg, optimizer)\u001b[0m\n\u001b[1;32m    642\u001b[0m     \u001b[38;5;66;03m# dnb_nll is the gradient wrt alpha, we need to multiply by alpha to get the\u001b[39;00m\n\u001b[1;32m    643\u001b[0m     \u001b[38;5;66;03m# gradient wrt log_alpha\u001b[39;00m\n\u001b[1;32m    644\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m alpha \u001b[38;5;241m*\u001b[39m dnb_nll(counts, mu, alpha) \u001b[38;5;241m+\u001b[39m reg_grad\n\u001b[0;32m--> 646\u001b[0m res \u001b[38;5;241m=\u001b[39m minimize(\n\u001b[1;32m    647\u001b[0m     \u001b[38;5;28;01mlambda\u001b[39;00m x: loss(x[\u001b[38;5;241m0\u001b[39m]),\n\u001b[1;32m    648\u001b[0m     x0\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mlog(alpha_hat),\n\u001b[1;32m    649\u001b[0m     jac\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mlambda\u001b[39;00m x: dloss(x[\u001b[38;5;241m0\u001b[39m]),\n\u001b[1;32m    650\u001b[0m     method\u001b[38;5;241m=\u001b[39moptimizer,\n\u001b[1;32m    651\u001b[0m     bounds\u001b[38;5;241m=\u001b[39m[(np\u001b[38;5;241m.\u001b[39mlog(min_disp), np\u001b[38;5;241m.\u001b[39mlog(max_disp))]\n\u001b[1;32m    652\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m optimizer \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mL-BFGS-B\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    653\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    654\u001b[0m )\n\u001b[1;32m    656\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m res\u001b[38;5;241m.\u001b[39msuccess:\n\u001b[1;32m    657\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39mexp(res\u001b[38;5;241m.\u001b[39mx[\u001b[38;5;241m0\u001b[39m]), res\u001b[38;5;241m.\u001b[39msuccess\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/scipy/optimize/_minimize.py:681\u001b[0m, in \u001b[0;36mminimize\u001b[0;34m(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)\u001b[0m\n\u001b[1;32m    677\u001b[0m                 \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcallable\u001b[39m(con\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mjac\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)):\n\u001b[1;32m    678\u001b[0m                     con[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mjac\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m _remove_from_func(con[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mjac\u001b[39m\u001b[38;5;124m'\u001b[39m], i_fixed,\n\u001b[1;32m    679\u001b[0m                                                    x_fixed, min_dim\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m,\n\u001b[1;32m    680\u001b[0m                                                    remove\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m--> 681\u001b[0m     bounds \u001b[38;5;241m=\u001b[39m standardize_bounds(bounds, x0, meth)\n\u001b[1;32m    683\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m meth \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnelder-mead\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    684\u001b[0m     res \u001b[38;5;241m=\u001b[39m _minimize_neldermead(fun, x0, args, callback, bounds\u001b[38;5;241m=\u001b[39mbounds,\n\u001b[1;32m    685\u001b[0m                                \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions)\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/scipy/optimize/_minimize.py:969\u001b[0m, in \u001b[0;36mstandardize_bounds\u001b[0;34m(bounds, x0, meth)\u001b[0m\n\u001b[1;32m    967\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m meth \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124ml-bfgs-b\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtnc\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mslsqp\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mold\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    968\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(bounds, Bounds):\n\u001b[0;32m--> 969\u001b[0m         bounds \u001b[38;5;241m=\u001b[39m new_bounds_to_old(bounds\u001b[38;5;241m.\u001b[39mlb, bounds\u001b[38;5;241m.\u001b[39mub, x0\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m])\n\u001b[1;32m    970\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m bounds\n",
      "File \u001b[0;32m~/anaconda3/envs/pydeseq2/lib/python3.11/site-packages/scipy/optimize/_constraints.py:398\u001b[0m, in \u001b[0;36mnew_bounds_to_old\u001b[0;34m(lb, ub, n)\u001b[0m\n\u001b[1;32m    389\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mnew_bounds_to_old\u001b[39m(lb, ub, n):\n\u001b[1;32m    390\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Convert the new bounds representation to the old one.\u001b[39;00m\n\u001b[1;32m    391\u001b[0m \n\u001b[1;32m    392\u001b[0m \u001b[38;5;124;03m    The new representation is a tuple (lb, ub) and the old one is a list\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    396\u001b[0m \u001b[38;5;124;03m    None.\u001b[39;00m\n\u001b[1;32m    397\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 398\u001b[0m     lb \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mbroadcast_to(lb, n)\n\u001b[1;32m    399\u001b[0m     ub \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mbroadcast_to(ub, n)\n\u001b[1;32m    401\u001b[0m     lb \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mfloat\u001b[39m(x) \u001b[38;5;28;01mif\u001b[39;00m x \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m-\u001b[39mnp\u001b[38;5;241m.\u001b[39minf \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m lb]\n",
      "File \u001b[0;32m<__array_function__ internals>:200\u001b[0m, in \u001b[0;36mbroadcast_to\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7441d4c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 6.18 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Annotation Tumor vs Normal\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_11</th>\n",
       "      <td>9.184301</td>\n",
       "      <td>2.523969</td>\n",
       "      <td>2.565003</td>\n",
       "      <td>0.984002</td>\n",
       "      <td>0.325114</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_14</th>\n",
       "      <td>2.024936</td>\n",
       "      <td>0.073297</td>\n",
       "      <td>2.349866</td>\n",
       "      <td>0.031192</td>\n",
       "      <td>0.975116</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_15</th>\n",
       "      <td>4.751814</td>\n",
       "      <td>-0.308316</td>\n",
       "      <td>1.574977</td>\n",
       "      <td>-0.195759</td>\n",
       "      <td>0.844799</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_16</th>\n",
       "      <td>2.598618</td>\n",
       "      <td>-1.382005</td>\n",
       "      <td>1.773673</td>\n",
       "      <td>-0.779177</td>\n",
       "      <td>0.435876</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_17</th>\n",
       "      <td>1.328844</td>\n",
       "      <td>1.880299</td>\n",
       "      <td>3.019965</td>\n",
       "      <td>0.622623</td>\n",
       "      <td>0.533532</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181882</th>\n",
       "      <td>3.560776</td>\n",
       "      <td>-0.957155</td>\n",
       "      <td>1.420471</td>\n",
       "      <td>-0.673829</td>\n",
       "      <td>0.500420</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181883</th>\n",
       "      <td>2.104789</td>\n",
       "      <td>1.550597</td>\n",
       "      <td>2.822595</td>\n",
       "      <td>0.549352</td>\n",
       "      <td>0.582764</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181884</th>\n",
       "      <td>2.505286</td>\n",
       "      <td>-0.741095</td>\n",
       "      <td>1.814857</td>\n",
       "      <td>-0.408349</td>\n",
       "      <td>0.683017</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181885</th>\n",
       "      <td>5.206586</td>\n",
       "      <td>-1.372438</td>\n",
       "      <td>1.237913</td>\n",
       "      <td>-1.108670</td>\n",
       "      <td>0.267572</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181886</th>\n",
       "      <td>12.449394</td>\n",
       "      <td>0.028422</td>\n",
       "      <td>1.074714</td>\n",
       "      <td>0.026446</td>\n",
       "      <td>0.978902</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175624 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "PeakID                                                                      \n",
       "cort_peak_11       9.184301        2.523969  2.565003  0.984002  0.325114   \n",
       "cort_peak_14       2.024936        0.073297  2.349866  0.031192  0.975116   \n",
       "cort_peak_15       4.751814       -0.308316  1.574977 -0.195759  0.844799   \n",
       "cort_peak_16       2.598618       -1.382005  1.773673 -0.779177  0.435876   \n",
       "cort_peak_17       1.328844        1.880299  3.019965  0.622623  0.533532   \n",
       "...                     ...             ...       ...       ...       ...   \n",
       "cort_peak_181882   3.560776       -0.957155  1.420471 -0.673829  0.500420   \n",
       "cort_peak_181883   2.104789        1.550597  2.822595  0.549352  0.582764   \n",
       "cort_peak_181884   2.505286       -0.741095  1.814857 -0.408349  0.683017   \n",
       "cort_peak_181885   5.206586       -1.372438  1.237913 -1.108670  0.267572   \n",
       "cort_peak_181886  12.449394        0.028422  1.074714  0.026446  0.978902   \n",
       "\n",
       "                      padj  \n",
       "PeakID                      \n",
       "cort_peak_11      0.999998  \n",
       "cort_peak_14      0.999998  \n",
       "cort_peak_15      0.999998  \n",
       "cort_peak_16      0.999998  \n",
       "cort_peak_17      0.999998  \n",
       "...                    ...  \n",
       "cort_peak_181882  0.999998  \n",
       "cort_peak_181883  0.999998  \n",
       "cort_peak_181884  0.999998  \n",
       "cort_peak_181885  0.999998  \n",
       "cort_peak_181886  0.999998  \n",
       "\n",
       "[175624 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_42326</th>\n",
       "      <td>28.625864</td>\n",
       "      <td>-2.419803</td>\n",
       "      <td>0.687452</td>\n",
       "      <td>-3.519960</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_16179</th>\n",
       "      <td>40.123711</td>\n",
       "      <td>2.754354</td>\n",
       "      <td>0.807089</td>\n",
       "      <td>3.412700</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_5971</th>\n",
       "      <td>8.813785</td>\n",
       "      <td>-3.369976</td>\n",
       "      <td>1.003237</td>\n",
       "      <td>-3.359101</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_11870</th>\n",
       "      <td>86.298444</td>\n",
       "      <td>2.266453</td>\n",
       "      <td>0.680046</td>\n",
       "      <td>3.332793</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_145189</th>\n",
       "      <td>35.197452</td>\n",
       "      <td>2.830571</td>\n",
       "      <td>0.863520</td>\n",
       "      <td>3.277945</td>\n",
       "      <td>0.001046</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_179586</th>\n",
       "      <td>3.413964</td>\n",
       "      <td>3.110540</td>\n",
       "      <td>4.663571</td>\n",
       "      <td>0.666987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_179897</th>\n",
       "      <td>1.573086</td>\n",
       "      <td>2.134140</td>\n",
       "      <td>3.319395</td>\n",
       "      <td>0.642930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_180546</th>\n",
       "      <td>1.788350</td>\n",
       "      <td>2.313598</td>\n",
       "      <td>3.244107</td>\n",
       "      <td>0.713169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_180661</th>\n",
       "      <td>2.701899</td>\n",
       "      <td>2.930165</td>\n",
       "      <td>2.796311</td>\n",
       "      <td>1.047868</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_181342</th>\n",
       "      <td>1.536985</td>\n",
       "      <td>0.624888</td>\n",
       "      <td>2.556146</td>\n",
       "      <td>0.244465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175624 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "PeakID                                                                      \n",
       "cort_peak_42326   28.625864       -2.419803  0.687452 -3.519960  0.000432   \n",
       "cort_peak_16179   40.123711        2.754354  0.807089  3.412700  0.000643   \n",
       "cort_peak_5971     8.813785       -3.369976  1.003237 -3.359101  0.000782   \n",
       "cort_peak_11870   86.298444        2.266453  0.680046  3.332793  0.000860   \n",
       "cort_peak_145189  35.197452        2.830571  0.863520  3.277945  0.001046   \n",
       "...                     ...             ...       ...       ...       ...   \n",
       "cort_peak_179586   3.413964        3.110540  4.663571  0.666987       NaN   \n",
       "cort_peak_179897   1.573086        2.134140  3.319395  0.642930       NaN   \n",
       "cort_peak_180546   1.788350        2.313598  3.244107  0.713169       NaN   \n",
       "cort_peak_180661   2.701899        2.930165  2.796311  1.047868       NaN   \n",
       "cort_peak_181342   1.536985        0.624888  2.556146  0.244465       NaN   \n",
       "\n",
       "                      padj  \n",
       "PeakID                      \n",
       "cort_peak_42326   0.999998  \n",
       "cort_peak_16179   0.999998  \n",
       "cort_peak_5971    0.999998  \n",
       "cort_peak_11870   0.999998  \n",
       "cort_peak_145189  0.999998  \n",
       "...                    ...  \n",
       "cort_peak_179586       NaN  \n",
       "cort_peak_179897       NaN  \n",
       "cort_peak_180546       NaN  \n",
       "cort_peak_180661       NaN  \n",
       "cort_peak_181342       NaN  \n",
       "\n",
       "[175624 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat_res = DeseqStats(dds, n_cpus=2)\n",
    "stat_res.summary()\n",
    "peak_results = stat_res.results_df\n",
    "peak_results = peak_results.sort_values('pvalue', ascending=True)\n",
    "peak_results.to_csv('../results/ATAC/differential/DAC_peaks.csv')\n",
    "peak_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2d793138",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6 × 19683\n",
       "    obs: 'Annotation'\n",
       "    obsm: 'design_matrix'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Genes\n",
    "dds = DeseqDataSet(\n",
    "    counts=gene_counts,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=\"Annotation\",\n",
    "    refit_cooks=True,\n",
    "    n_cpus=1,\n",
    ")\n",
    "dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3e9cb1f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 5.61 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 1.45 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 5.60 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.52 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "dd0a0d58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 1.79 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Annotation Tumor vs Normal\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>99.692270</td>\n",
       "      <td>0.440064</td>\n",
       "      <td>0.478212</td>\n",
       "      <td>0.920227</td>\n",
       "      <td>0.357454</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>33.937291</td>\n",
       "      <td>-0.301201</td>\n",
       "      <td>0.628244</td>\n",
       "      <td>-0.479433</td>\n",
       "      <td>0.631631</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>280.865414</td>\n",
       "      <td>0.361631</td>\n",
       "      <td>0.333818</td>\n",
       "      <td>1.083318</td>\n",
       "      <td>0.278667</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>288.181869</td>\n",
       "      <td>0.626359</td>\n",
       "      <td>0.328406</td>\n",
       "      <td>1.907273</td>\n",
       "      <td>0.056485</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>327.849676</td>\n",
       "      <td>-0.392905</td>\n",
       "      <td>0.374722</td>\n",
       "      <td>-1.048524</td>\n",
       "      <td>0.294397</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-210D15.9</th>\n",
       "      <td>10.779070</td>\n",
       "      <td>-0.444944</td>\n",
       "      <td>0.950392</td>\n",
       "      <td>-0.468169</td>\n",
       "      <td>0.639664</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-244E17.1</th>\n",
       "      <td>362.866479</td>\n",
       "      <td>0.278642</td>\n",
       "      <td>0.332496</td>\n",
       "      <td>0.838033</td>\n",
       "      <td>0.402012</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTB-60B18.23</th>\n",
       "      <td>94.591196</td>\n",
       "      <td>-0.107540</td>\n",
       "      <td>0.431677</td>\n",
       "      <td>-0.249122</td>\n",
       "      <td>0.803267</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP3-454G6.2</th>\n",
       "      <td>98.269038</td>\n",
       "      <td>0.055708</td>\n",
       "      <td>0.476292</td>\n",
       "      <td>0.116962</td>\n",
       "      <td>0.906890</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP5-937E21.8</th>\n",
       "      <td>6.677200</td>\n",
       "      <td>0.354412</td>\n",
       "      <td>1.357742</td>\n",
       "      <td>0.261030</td>\n",
       "      <td>0.794069</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19683 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "TSPAN6          99.692270        0.440064  0.478212  0.920227  0.357454   \n",
       "TNMD            33.937291       -0.301201  0.628244 -0.479433  0.631631   \n",
       "DPM1           280.865414        0.361631  0.333818  1.083318  0.278667   \n",
       "SCYL3          288.181869        0.626359  0.328406  1.907273  0.056485   \n",
       "C1orf112       327.849676       -0.392905  0.374722 -1.048524  0.294397   \n",
       "...                   ...             ...       ...       ...       ...   \n",
       "RP13-210D15.9   10.779070       -0.444944  0.950392 -0.468169  0.639664   \n",
       "RP11-244E17.1  362.866479        0.278642  0.332496  0.838033  0.402012   \n",
       "CTB-60B18.23    94.591196       -0.107540  0.431677 -0.249122  0.803267   \n",
       "RP3-454G6.2     98.269038        0.055708  0.476292  0.116962  0.906890   \n",
       "RP5-937E21.8     6.677200        0.354412  1.357742  0.261030  0.794069   \n",
       "\n",
       "                   padj  \n",
       "TSPAN6         0.997348  \n",
       "TNMD           0.997348  \n",
       "DPM1           0.997348  \n",
       "SCYL3          0.997348  \n",
       "C1orf112       0.997348  \n",
       "...                 ...  \n",
       "RP13-210D15.9  0.997348  \n",
       "RP11-244E17.1  0.997348  \n",
       "CTB-60B18.23   0.997348  \n",
       "RP3-454G6.2    0.997348  \n",
       "RP5-937E21.8   0.997348  \n",
       "\n",
       "[19683 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LRFN5</th>\n",
       "      <td>332.880470</td>\n",
       "      <td>-1.323594</td>\n",
       "      <td>0.406974</td>\n",
       "      <td>-3.252278</td>\n",
       "      <td>0.001145</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRYBG3</th>\n",
       "      <td>259.833503</td>\n",
       "      <td>-1.209481</td>\n",
       "      <td>0.376936</td>\n",
       "      <td>-3.208717</td>\n",
       "      <td>0.001333</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGCS2</th>\n",
       "      <td>46.137177</td>\n",
       "      <td>-1.660528</td>\n",
       "      <td>0.530240</td>\n",
       "      <td>-3.131655</td>\n",
       "      <td>0.001738</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRAC2</th>\n",
       "      <td>37.371114</td>\n",
       "      <td>-1.860092</td>\n",
       "      <td>0.598162</td>\n",
       "      <td>-3.109677</td>\n",
       "      <td>0.001873</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP9</th>\n",
       "      <td>75.474269</td>\n",
       "      <td>-1.478217</td>\n",
       "      <td>0.483820</td>\n",
       "      <td>-3.055306</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2CD4C</th>\n",
       "      <td>81.861119</td>\n",
       "      <td>-0.000108</td>\n",
       "      <td>0.442278</td>\n",
       "      <td>-0.000245</td>\n",
       "      <td>0.999805</td>\n",
       "      <td>0.999891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MED23</th>\n",
       "      <td>212.733422</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.323466</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.999859</td>\n",
       "      <td>0.999891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOPORS-AS1</th>\n",
       "      <td>274.730846</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.359993</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.999891</td>\n",
       "      <td>0.999891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C21orf33</th>\n",
       "      <td>4.777863</td>\n",
       "      <td>3.667897</td>\n",
       "      <td>4.652546</td>\n",
       "      <td>0.788363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PWP2</th>\n",
       "      <td>6.052068</td>\n",
       "      <td>4.004196</td>\n",
       "      <td>4.650456</td>\n",
       "      <td>0.861033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19683 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "LRFN5       332.880470       -1.323594  0.406974 -3.252278  0.001145  0.997348\n",
       "CRYBG3      259.833503       -1.209481  0.376936 -3.208717  0.001333  0.997348\n",
       "HMGCS2       46.137177       -1.660528  0.530240 -3.131655  0.001738  0.997348\n",
       "PRAC2        37.371114       -1.860092  0.598162 -3.109677  0.001873  0.997348\n",
       "AQP9         75.474269       -1.478217  0.483820 -3.055306  0.002248  0.997348\n",
       "...                ...             ...       ...       ...       ...       ...\n",
       "C2CD4C       81.861119       -0.000108  0.442278 -0.000245  0.999805  0.999891\n",
       "MED23       212.733422        0.000057  0.323466  0.000177  0.999859  0.999891\n",
       "TOPORS-AS1  274.730846        0.000049  0.359993  0.000137  0.999891  0.999891\n",
       "C21orf33      4.777863        3.667897  4.652546  0.788363       NaN       NaN\n",
       "PWP2          6.052068        4.004196  4.650456  0.861033       NaN       NaN\n",
       "\n",
       "[19683 rows x 6 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat_res = DeseqStats(dds, n_cpus=2)\n",
    "stat_res.summary()\n",
    "gene_results = stat_res.results_df\n",
    "gene_results = gene_results.sort_values('pvalue', ascending=True)\n",
    "gene_results.to_csv('../results/ATAC/differential/DAC_gene.csv')\n",
    "gene_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2ad3d80b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LRFN5</th>\n",
       "      <td>332.880470</td>\n",
       "      <td>-1.323594</td>\n",
       "      <td>0.406974</td>\n",
       "      <td>-3.252278</td>\n",
       "      <td>0.001145</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRYBG3</th>\n",
       "      <td>259.833503</td>\n",
       "      <td>-1.209481</td>\n",
       "      <td>0.376936</td>\n",
       "      <td>-3.208717</td>\n",
       "      <td>0.001333</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGCS2</th>\n",
       "      <td>46.137177</td>\n",
       "      <td>-1.660528</td>\n",
       "      <td>0.530240</td>\n",
       "      <td>-3.131655</td>\n",
       "      <td>0.001738</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRAC2</th>\n",
       "      <td>37.371114</td>\n",
       "      <td>-1.860092</td>\n",
       "      <td>0.598162</td>\n",
       "      <td>-3.109677</td>\n",
       "      <td>0.001873</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP9</th>\n",
       "      <td>75.474269</td>\n",
       "      <td>-1.478217</td>\n",
       "      <td>0.483820</td>\n",
       "      <td>-3.055306</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MID1IP1</th>\n",
       "      <td>151.368361</td>\n",
       "      <td>0.732034</td>\n",
       "      <td>0.373020</td>\n",
       "      <td>1.962451</td>\n",
       "      <td>0.049710</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH19</th>\n",
       "      <td>120.434175</td>\n",
       "      <td>-0.938305</td>\n",
       "      <td>0.478328</td>\n",
       "      <td>-1.961635</td>\n",
       "      <td>0.049805</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARRDC5</th>\n",
       "      <td>46.543657</td>\n",
       "      <td>-1.013628</td>\n",
       "      <td>0.516953</td>\n",
       "      <td>-1.960774</td>\n",
       "      <td>0.049905</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANXA1</th>\n",
       "      <td>58.475922</td>\n",
       "      <td>-1.177671</td>\n",
       "      <td>0.600673</td>\n",
       "      <td>-1.960585</td>\n",
       "      <td>0.049927</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SARDH</th>\n",
       "      <td>586.763817</td>\n",
       "      <td>-0.585131</td>\n",
       "      <td>0.298458</td>\n",
       "      <td>-1.960518</td>\n",
       "      <td>0.049935</td>\n",
       "      <td>0.997348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>517 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "LRFN5    332.880470       -1.323594  0.406974 -3.252278  0.001145  0.997348\n",
       "CRYBG3   259.833503       -1.209481  0.376936 -3.208717  0.001333  0.997348\n",
       "HMGCS2    46.137177       -1.660528  0.530240 -3.131655  0.001738  0.997348\n",
       "PRAC2     37.371114       -1.860092  0.598162 -3.109677  0.001873  0.997348\n",
       "AQP9      75.474269       -1.478217  0.483820 -3.055306  0.002248  0.997348\n",
       "...             ...             ...       ...       ...       ...       ...\n",
       "MID1IP1  151.368361        0.732034  0.373020  1.962451  0.049710  0.997348\n",
       "CDH19    120.434175       -0.938305  0.478328 -1.961635  0.049805  0.997348\n",
       "ARRDC5    46.543657       -1.013628  0.516953 -1.960774  0.049905  0.997348\n",
       "ANXA1     58.475922       -1.177671  0.600673 -1.960585  0.049927  0.997348\n",
       "SARDH    586.763817       -0.585131  0.298458 -1.960518  0.049935  0.997348\n",
       "\n",
       "[517 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# peak_results[peak_results['pvalue'] < 0.05]#['log2FoldChange'] > 0\n",
    "gene_results[gene_results['pvalue'] < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "f6c23d4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PeakID</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cort_peak_42326</td>\n",
       "      <td>28.625864</td>\n",
       "      <td>-2.419803</td>\n",
       "      <td>0.687452</td>\n",
       "      <td>-3.519960</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cort_peak_16179</td>\n",
       "      <td>40.123711</td>\n",
       "      <td>2.754354</td>\n",
       "      <td>0.807089</td>\n",
       "      <td>3.412700</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cort_peak_5971</td>\n",
       "      <td>8.813785</td>\n",
       "      <td>-3.369976</td>\n",
       "      <td>1.003237</td>\n",
       "      <td>-3.359101</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cort_peak_11870</td>\n",
       "      <td>86.298444</td>\n",
       "      <td>2.266453</td>\n",
       "      <td>0.680046</td>\n",
       "      <td>3.332793</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cort_peak_145189</td>\n",
       "      <td>35.197452</td>\n",
       "      <td>2.830571</td>\n",
       "      <td>0.863520</td>\n",
       "      <td>3.277945</td>\n",
       "      <td>0.001046</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175619</th>\n",
       "      <td>cort_peak_179586</td>\n",
       "      <td>3.413964</td>\n",
       "      <td>3.110540</td>\n",
       "      <td>4.663571</td>\n",
       "      <td>0.666987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175620</th>\n",
       "      <td>cort_peak_179897</td>\n",
       "      <td>1.573086</td>\n",
       "      <td>2.134140</td>\n",
       "      <td>3.319395</td>\n",
       "      <td>0.642930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175621</th>\n",
       "      <td>cort_peak_180546</td>\n",
       "      <td>1.788350</td>\n",
       "      <td>2.313598</td>\n",
       "      <td>3.244107</td>\n",
       "      <td>0.713169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175622</th>\n",
       "      <td>cort_peak_180661</td>\n",
       "      <td>2.701899</td>\n",
       "      <td>2.930165</td>\n",
       "      <td>2.796311</td>\n",
       "      <td>1.047868</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175623</th>\n",
       "      <td>cort_peak_181342</td>\n",
       "      <td>1.536985</td>\n",
       "      <td>0.624888</td>\n",
       "      <td>2.556146</td>\n",
       "      <td>0.244465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175624 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  PeakID   baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0        cort_peak_42326  28.625864       -2.419803  0.687452 -3.519960   \n",
       "1        cort_peak_16179  40.123711        2.754354  0.807089  3.412700   \n",
       "2         cort_peak_5971   8.813785       -3.369976  1.003237 -3.359101   \n",
       "3        cort_peak_11870  86.298444        2.266453  0.680046  3.332793   \n",
       "4       cort_peak_145189  35.197452        2.830571  0.863520  3.277945   \n",
       "...                  ...        ...             ...       ...       ...   \n",
       "175619  cort_peak_179586   3.413964        3.110540  4.663571  0.666987   \n",
       "175620  cort_peak_179897   1.573086        2.134140  3.319395  0.642930   \n",
       "175621  cort_peak_180546   1.788350        2.313598  3.244107  0.713169   \n",
       "175622  cort_peak_180661   2.701899        2.930165  2.796311  1.047868   \n",
       "175623  cort_peak_181342   1.536985        0.624888  2.556146  0.244465   \n",
       "\n",
       "          pvalue      padj  \n",
       "0       0.000432  0.999998  \n",
       "1       0.000643  0.999998  \n",
       "2       0.000782  0.999998  \n",
       "3       0.000860  0.999998  \n",
       "4       0.001046  0.999998  \n",
       "...          ...       ...  \n",
       "175619       NaN       NaN  \n",
       "175620       NaN       NaN  \n",
       "175621       NaN       NaN  \n",
       "175622       NaN       NaN  \n",
       "175623       NaN       NaN  \n",
       "\n",
       "[175624 rows x 7 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## generate peak files for homer analysis\n",
    "peak_results = pd.read_csv('../results/ATAC/differential/DAC_peaks.csv')\n",
    "# peak_results.index = peak_results['PeakID']\n",
    "# peak_results = peak_results.drop(['PeakID'], axis=1)\n",
    "peak_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2e27cbbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Peak Score</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Distance to TSS</th>\n",
       "      <th>Nearest Ensembl</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_72682</th>\n",
       "      <td>chr17</td>\n",
       "      <td>53105799</td>\n",
       "      <td>53106413</td>\n",
       "      <td>+</td>\n",
       "      <td>1760</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>120601.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C17orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_116019</th>\n",
       "      <td>chr3</td>\n",
       "      <td>93470333</td>\n",
       "      <td>93470821</td>\n",
       "      <td>+</td>\n",
       "      <td>1648</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>503504.0</td>\n",
       "      <td>ENSG00000184500</td>\n",
       "      <td>PROS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_49208</th>\n",
       "      <td>chr13</td>\n",
       "      <td>109424086</td>\n",
       "      <td>109424398</td>\n",
       "      <td>+</td>\n",
       "      <td>1567</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>23546.0</td>\n",
       "      <td>ENSG00000229792</td>\n",
       "      <td>LINC00399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_55823</th>\n",
       "      <td>chr15</td>\n",
       "      <td>20505961</td>\n",
       "      <td>20506672</td>\n",
       "      <td>+</td>\n",
       "      <td>1194</td>\n",
       "      <td>promoter-TSS (NR_036432)</td>\n",
       "      <td>-136.0</td>\n",
       "      <td>ENSG00000180229</td>\n",
       "      <td>HERC2P3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_82351</th>\n",
       "      <td>chr19</td>\n",
       "      <td>20999554</td>\n",
       "      <td>21000284</td>\n",
       "      <td>+</td>\n",
       "      <td>1071</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>-20738.0</td>\n",
       "      <td>ENSG00000118620</td>\n",
       "      <td>ZNF430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_167501</th>\n",
       "      <td>chr8</td>\n",
       "      <td>127823141</td>\n",
       "      <td>127823257</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>intron (NR_003367, intron 1 of 8)</td>\n",
       "      <td>27237.0</td>\n",
       "      <td>ENSG00000283710</td>\n",
       "      <td>MIR1204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_95046</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178065570</td>\n",
       "      <td>178065783</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>intron (NM_001077197, intron 2 of 20)</td>\n",
       "      <td>7101.0</td>\n",
       "      <td>ENSG00000128655</td>\n",
       "      <td>PDE11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_148764</th>\n",
       "      <td>chr6</td>\n",
       "      <td>139425413</td>\n",
       "      <td>139425558</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>49111.0</td>\n",
       "      <td>ENSG00000238099</td>\n",
       "      <td>LINC01625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_134843</th>\n",
       "      <td>chr5</td>\n",
       "      <td>94573465</td>\n",
       "      <td>94573611</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>intron (NM_001145678, intron 2 of 20)</td>\n",
       "      <td>45066.0</td>\n",
       "      <td>ENSG00000185261</td>\n",
       "      <td>KIAA0825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_2365</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11690774</td>\n",
       "      <td>11690897</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>promoter-TSS (NM_198545)</td>\n",
       "      <td>786.0</td>\n",
       "      <td>ENSG00000116670</td>\n",
       "      <td>MAD2L2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181886 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Chr      Start        End Strand  Peak Score  \\\n",
       "PeakID                                                             \n",
       "cort_peak_72682   chr17   53105799   53106413      +        1760   \n",
       "cort_peak_116019   chr3   93470333   93470821      +        1648   \n",
       "cort_peak_49208   chr13  109424086  109424398      +        1567   \n",
       "cort_peak_55823   chr15   20505961   20506672      +        1194   \n",
       "cort_peak_82351   chr19   20999554   21000284      +        1071   \n",
       "...                 ...        ...        ...    ...         ...   \n",
       "cort_peak_167501   chr8  127823141  127823257      +          13   \n",
       "cort_peak_95046    chr2  178065570  178065783      +          13   \n",
       "cort_peak_148764   chr6  139425413  139425558      +          13   \n",
       "cort_peak_134843   chr5   94573465   94573611      +          13   \n",
       "cort_peak_2365     chr1   11690774   11690897      +          13   \n",
       "\n",
       "                                             Annotation  Distance to TSS  \\\n",
       "PeakID                                                                     \n",
       "cort_peak_72682                              Intergenic         120601.0   \n",
       "cort_peak_116019                             Intergenic         503504.0   \n",
       "cort_peak_49208                              Intergenic          23546.0   \n",
       "cort_peak_55823                promoter-TSS (NR_036432)           -136.0   \n",
       "cort_peak_82351                              Intergenic         -20738.0   \n",
       "...                                                 ...              ...   \n",
       "cort_peak_167501      intron (NR_003367, intron 1 of 8)          27237.0   \n",
       "cort_peak_95046   intron (NM_001077197, intron 2 of 20)           7101.0   \n",
       "cort_peak_148764                             Intergenic          49111.0   \n",
       "cort_peak_134843  intron (NM_001145678, intron 2 of 20)          45066.0   \n",
       "cort_peak_2365                 promoter-TSS (NM_198545)            786.0   \n",
       "\n",
       "                  Nearest Ensembl  Gene Name  \n",
       "PeakID                                        \n",
       "cort_peak_72682               NaN  C17orf112  \n",
       "cort_peak_116019  ENSG00000184500      PROS1  \n",
       "cort_peak_49208   ENSG00000229792  LINC00399  \n",
       "cort_peak_55823   ENSG00000180229    HERC2P3  \n",
       "cort_peak_82351   ENSG00000118620     ZNF430  \n",
       "...                           ...        ...  \n",
       "cort_peak_167501  ENSG00000283710    MIR1204  \n",
       "cort_peak_95046   ENSG00000128655     PDE11A  \n",
       "cort_peak_148764  ENSG00000238099  LINC01625  \n",
       "cort_peak_134843  ENSG00000185261   KIAA0825  \n",
       "cort_peak_2365    ENSG00000116670     MAD2L2  \n",
       "\n",
       "[181886 rows x 9 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_annotated = pd.read_csv('../results/ATAC/macs/cort_peaks_annotated.narrowPeak', sep='\\t')\n",
    "peaks_annotated = peaks_annotated[['PeakID (cmd=annotatePeaks.pl results/ATAC/macs/cort_peaks.narrowPeak hg38)','Chr','Start','End','Strand','Peak Score','Annotation','Distance to TSS','Nearest Ensembl','Gene Name']]\n",
    "peaks_annotated.columns = ['PeakID','Chr','Start','End','Strand','Peak Score','Annotation','Distance to TSS','Nearest Ensembl','Gene Name']\n",
    "# peaks.index = peaks['PeakID']\n",
    "# peaks = peaks.drop(['PeakID'], axis=1)\n",
    "peaks_annotated.index = peaks_annotated['PeakID']\n",
    "peaks_annotated = peaks_annotated.drop(['PeakID'], axis=1)\n",
    "peaks_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "30d2e309",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_results_annotated = pd.DataFrame(peak_results)\n",
    "peak_results_annotated.index = peak_results['PeakID']\n",
    "peak_results_annotated[['Chr','Start','End','Strand','Peak Score','Annotation','Distance to TSS','Nearest Ensembl','Gene Name']] = peak_results_annotated.apply(lambda x: peaks_annotated.loc[x['PeakID']], axis=1, result_type='expand')\n",
    "peak_results.index = peak_results['PeakID']\n",
    "peak_results = peak_results.drop(['PeakID'], axis=1)\n",
    "peak_results_annotated.to_csv('../results/ATAC/differential/DAC_peaks_annotated.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "8a02c42f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cort_peak_16179</th>\n",
       "      <td>chr1</td>\n",
       "      <td>221839455</td>\n",
       "      <td>221840294</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_11870</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161242446</td>\n",
       "      <td>161243287</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_145189</th>\n",
       "      <td>chr6</td>\n",
       "      <td>75301970</td>\n",
       "      <td>75302682</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_107633</th>\n",
       "      <td>chr22</td>\n",
       "      <td>26132928</td>\n",
       "      <td>26133566</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_71236</th>\n",
       "      <td>chr17</td>\n",
       "      <td>40417608</td>\n",
       "      <td>40418448</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_177661</th>\n",
       "      <td>chrX</td>\n",
       "      <td>20098824</td>\n",
       "      <td>20099209</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_61310</th>\n",
       "      <td>chr15</td>\n",
       "      <td>92625196</td>\n",
       "      <td>92625906</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_31525</th>\n",
       "      <td>chr11</td>\n",
       "      <td>68030091</td>\n",
       "      <td>68031087</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_113429</th>\n",
       "      <td>chr3</td>\n",
       "      <td>46646493</td>\n",
       "      <td>46647091</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cort_peak_143841</th>\n",
       "      <td>chr6</td>\n",
       "      <td>41936245</td>\n",
       "      <td>41936717</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>606 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Chr      Start        End Strand\n",
       "PeakID                                              \n",
       "cort_peak_16179    chr1  221839455  221840294      +\n",
       "cort_peak_11870    chr1  161242446  161243287      +\n",
       "cort_peak_145189   chr6   75301970   75302682      +\n",
       "cort_peak_107633  chr22   26132928   26133566      +\n",
       "cort_peak_71236   chr17   40417608   40418448      +\n",
       "...                 ...        ...        ...    ...\n",
       "cort_peak_177661   chrX   20098824   20099209      +\n",
       "cort_peak_61310   chr15   92625196   92625906      +\n",
       "cort_peak_31525   chr11   68030091   68031087      +\n",
       "cort_peak_113429   chr3   46646493   46647091      +\n",
       "cort_peak_143841   chr6   41936245   41936717      +\n",
       "\n",
       "[606 rows x 4 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig = peak_results_annotated[peak_results_annotated['pvalue']<0.05]\n",
    "homer_up = sig[sig['log2FoldChange']>0.5][['Chr','Start','End','Strand']]\n",
    "homer_up.to_csv('../results/ATAC/homer/peaks_up.homer', sep='\\t')\n",
    "homer_down = sig[sig['log2FoldChange']<-0.5][['Chr','Start','End','Strand']]\n",
    "homer_down.to_csv('../results/ATAC/homer/peaks_down.homer', sep='\\t')\n",
    "homer_up"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
